common/tb.go

```diff
@@ -185,3 +211,79 @@ func (tb *tokenBucketImpl) isRefillDue(now int64) bool {
 func (tb *tokenBucketImpl) isOverflowRefillDue(now int64) bool {
 	return now >= tb.nextOverflowRefillTime
 }
+
+// NewPriorityTokenBucket creates and returns a
+// new token bucket rate limiter support priority.
+// There are n buckets for n priorities. It
+// replenishes the top priority bucket every 100
+// milliseconds, unused tokens flows to next bucket.
+// The idea comes from Dual Token Bucket Algorithms.
+// Thread safe.
+//
+// @param numOfPriority
+//    Number of priorities
+// @param rps
+//    Desired rate per second
+//
+func NewPriorityTokenBucket(numOfPriority, rps int, timeSource TimeSource) PriorityTokenBucket {
+	tb := new(priorityTokenBucketImpl)
+	tb.tokens = make([]int, numOfPriority)
+	tb.timeSource = timeSource
+	tb.fillInterval = int64(time.Millisecond * 100)
+	tb.fillRate = (rps * 100) / millisPerSecond
+	tb.overflowRps = rps - (10 * tb.fillRate)
+	tb.refill(time.Now().UnixNano())
+	return tb
+}
+
+func (tb *priorityTokenBucketImpl) GetToken(priority, count int) (bool, time.Duration) {
+	now := tb.timeSource.Now().UnixNano()
+	tb.Lock()
+	defer tb.Unlock()
+	tb.refill(now)
+	nextRefillTime := time.Duration(tb.nextRefillTime - now)
+	if tb.tokens[priority] < count {
+		return false, nextRefillTime
+	}
+	tb.tokens[priority] -= count
+	return true, nextRefillTime
+}
+
+func (tb *priorityTokenBucketImpl) refill(now int64) {
+	tb.refillOverFlow(now)
+	if tb.isRefillDue(now) {
+		more := tb.fillRate
```

Comment:
this `more` is the total tokens to be refilled for all level of priority.
so it would be good to rename this to remainingTokens or totalTokens
and add a comment here explaining the behavior

Reply:
It's the additional tokens for each buckets, none of these names looks perfect so far, so will keep it simple as more for now.